File: java/vm/src/test/java/simplevm/vm/StackTests.java
package simplevm.vm;

import org.junit.jupiter.api.Test;

import static org.junit.jupiter.api.Assertions.assertEquals;
import static org.junit.jupiter.api.Assertions.assertTrue;

import static simplevm.vm.Bytecode.*;

public class StackTests {
    @Test void testPushPop() {
        VirtualMachine vm = new VirtualMachine();

        vm.push(27);

        assertEquals(1, vm.getStack().length);
        assertEquals(27, vm.getStack()[0]);

        int value = vm.pop();
        assertEquals(27, value);
        assertEquals(0, vm.getStack().length);
    }
    @Test void TestCONST()
    {
        VirtualMachine vm = new VirtualMachine();

        vm.execute(CONST, 34);
        vm.execute(new int[] {
            CONST, 27,
        });

        assertEquals(2, vm.getStack().length);
        assertEquals(34, vm.getStack()[0]);
        assertEquals(27, vm.getStack()[1]);
    }
    @Test void TestCONSTPOP()
    {
        VirtualMachine vm = new VirtualMachine();

        vm.execute(new int[] {
            CONST, 27,
            POP,
        });

        assertEquals(0, vm.getStack().length);
    }
}

File: cs/vmtest/StackTests.cs
namespace vmtest;

using vm;

[TestClass]
public class StackTests
{
    [TestMethod]
    public void TestPush()
    {
        VirtualMachine vm = new VirtualMachine();

        vm.Push(27);

        Assert.AreEqual(1, vm.Stack.Length);
        Assert.AreEqual(27, vm.Stack[0]);
    }
    [TestMethod]
    public void TestPushPop()
    {
        VirtualMachine vm = new VirtualMachine();

        vm.Push(27);
        Assert.AreEqual(1, vm.Stack.Length);

        int value = vm.Pop();

        Assert.AreEqual(27, value);
        Assert.AreEqual(0, vm.Stack.Length);
    }
    [TestMethod]
    public void TestCONST()
    {
        VirtualMachine vm = new VirtualMachine();

        vm.Execute(Bytecode.CONST, 34);
        vm.Execute(new Bytecode[] {
            Bytecode.CONST, (Bytecode)27,
        });

        Assert.AreEqual(2, vm.Stack.Length);
        Assert.AreEqual(34, vm.Stack[0]);
        Assert.AreEqual(27, vm.Stack[1]);
    }
    [TestMethod]
    public void TestCONSTPOP()
    {
        VirtualMachine vm = new VirtualMachine();

        vm.Execute(new Bytecode[] {
            Bytecode.CONST, (Bytecode)27,
            Bytecode.POP,
        });

        Assert.AreEqual(0, vm.Stack.Length);
    }
}
